django/db/backends/base/schema.py

```diff
@@ -767,19 +726,70 @@ def _alter_field(self, model, old_field, new_field, old_type, new_type,
         # Drop the default if we need to
         # (Django usually does not use in-database defaults)
         if needs_database_default:
+            sql, params = self._alter_column_default_sql(model, old_field, new_field, True)
             sql = self.sql_alter_column % {
                 "table": self.quote_name(model._meta.db_table),
-                "changes": self.sql_alter_column_no_default % {
-                    "column": self.quote_name(new_field.column),
-                    "type": new_type,
-                }
+                "changes": sql,
             }
-            self.execute(sql)
+            self.execute(sql, params)
         # Reset connection if required
         if self.connection.features.connection_persists_old_columns:
             self.connection.close()
 
-    def _alter_column_type_sql(self, table, old_field, new_field, new_type):
+    def _alter_column_null_sql(self, model, old_field, new_field):
+        """
+        Hook to specialize column null alteration for different backends.
+
+        Return a sql fragment (sql, params).
+        """
+        new_db_params = new_field.db_parameters(connection=self.connection)
+
+        if (self.connection.features.interprets_empty_strings_as_nulls and
+                new_field.get_internal_type() in ("CharField", "TextField")):
+            # The field is nullable in the database anyway, leave it alone
+            return
+        else:
+            sql = self.sql_alter_column_null if new_field.null else self.sql_alter_column_not_null
+            return (
+                sql % {
+                    "column": self.quote_name(new_field.column),
+                    "type": new_db_params['type'],
+                },
+                [],
+            )
+
+    def _alter_column_default_sql(self, model, old_field, new_field, drop=False):
+        """
+        Hook to specialize column default alteration for different backends.
```

Comment:
I'd say, "Return a (sql, params) fragment to add or drop (depending on the drop argument) a default to new_field's column."